apt/message.go

```diff
@@ -0,0 +1,258 @@
+}
+
+// FailURI writes a 400 URI Failure message.
+func (w *AptMessageWriter) FailURI(uri, msg string) error {
+	return w.WriteMessage(new400Message(uri, msg))
+}
+
+// Fail writes a 401 General Failure message.
+func (w *AptMessageWriter) Fail(msg string) error {
+	return w.WriteMessage(new401Message(msg))
+}
+
+// ReadMessage reads lines from `reader` until a complete message is received.
+func (r *AptMessageReader) ReadMessage(ctx context.Context) (*AptMessage, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+		line, err := r.reader.ReadString('\n')
+		if err != nil {
+			/*
+				if err == io.EOF || err == io.ErrClosedPipe {
+					// TODO: what to return in this case?
+				}
+			*/
+			return nil, err
+		}
+
+		line = strings.TrimSpace(line)
+		if line == "" {
+			if r.message == nil {
+				return nil, errors.New("Empty message")
+			}
+
+			// Message is done, return and reset.
+			msg := r.message
+			r.message = nil
+			return msg, nil
+		}
+
+		if r.message == nil {
+			r.message = &AptMessage{}
+			if err := r.parseHeader(line); err != nil {
+				return nil, err
+			}
+		} else {
+			if err := r.parseField(line); err != nil {
+				return nil, err
+			}
+		}
+	}
+}
+
+func (r *AptMessageReader) parseHeader(line string) error {
+	if line == "" {
+		return errors.New("Empty message header")
+	}
+	if r.message.code != 0 || r.message.description != "" {
+		return errors.New("Double parsing header")
+	}
+	line = strings.TrimSpace(line)
+	parts := strings.SplitN(line, " ", 2)
+	if len(parts) != 2 {
+		return errors.New("Malformed header")
+	}
+	code, err := strconv.Atoi(strings.TrimSpace(parts[0]))
+	if err != nil {
+		return errors.New("Malformed header")
+	}
+
+	r.message.code = code
+	r.message.description = strings.TrimSpace(parts[1])
+	return nil
+}
+
+func (r *AptMessageReader) parseField(line string) error {
+	if line == "" {
+		return errors.New("Empty message field")
+	}
+	line = strings.TrimSpace(line)
+	parts := strings.SplitN(line, ":", 2)
+	if len(parts) < 2 {
+		return errors.New("Malformed field")
+	}
+	if r.message.fields == nil {
+		r.message.fields = make(map[string][]string)
+	}
+	key := strings.TrimSpace(parts[0])
+	value := strings.TrimSpace(parts[1])
+	if key == "" || value == "" {
+		return errors.New("Malformed field")
+	}
+
+	fieldlist := r.message.fields[key]
+	fieldlist = append(fieldlist, value)
+	r.message.fields[key] = fieldlist
+	return nil
+}
+
+func new100Message() AptMessage {
+	fields := make(map[string][]string)
+	fields["Send-Config"] = []string{"true"}
+	fields["Version"] = []string{"1.0"}
+	return AptMessage{code: 100, description: "Capabilities", fields: fields}
+}
+
+func new101Message(msg string) AptMessage {
+	fields := make(map[string][]string)
+	fields["Message"] = []string{msg}
+	return AptMessage{code: 101, description: "Log", fields: fields}
+}
+
+func new200Message(uri, size, lastModified string) AptMessage {
```

Comment:
yes, since we don't care about the size ourselves, there's no reason to bother parsing it out. as far as we are concerned these are opaque values we are passing from the http header (rfc822, uses string) to the apt message field (rfc822, uses string)